Upgraded to Rails v7

yedhink opened the pull request:
Fixes #783 

ActiveAdmin won't work: https://github.com/bigbinary/wheel/issues/783#issuecomment-1041805682

Will take up some other goodies integration in separate issues. Like:
- https://github.com/bigbinary/wheel/issues/785
- https://github.com/bigbinary/wheel/issues/786

Expecting deployment issues in Heroku. Will wait for deployment to complete before assigning to reviewer

yedhink commented:
@chiraggshah _a Can you please review this PR? Will tag Unni once this PR is good to be merged. And then we can take up syncing neeto projects via compliance.

cc: @ghousemohamed

chiraggshah reviewed (commented):
@yedhink _a : Added a few comments.

yedhink commented on config/application.rb:
```diff
@@ -1,5 +1,3 @@
-# frozen_string_literal: true
```
@chiraggshah I have disabled Rubocop for `config` and it's subfolders because either ways configs will get modified during upgrade process with Rails. Didn't think Rubocop will add any value additions for `config` folder.

Lemme know your thoughts on the same.

yedhink replied:
Have updated the rules such that config folder is under Rubocop scope but it won't enforce frozen string literal rule for `config` and sub-directories. Frozen string literal gets removed by Rails under config folder whenever we upgrade. And I don't think adding this comment for `config` folder doesn't add much value.

yedhink commented on .rubocop.yml:
```diff
@@ -434,13 +435,13 @@ Layout/LineLength:
+      '"$', # line that ends with double quote (long string variable assignment)
+      "'$", # line that ends with single quote (long string variable assignment)
+      '\/$', # line that ends with slash (long regex variable assignment)
+    ]
```
This was unintentional. This formatting is being done by the vscode-yaml extension. And it's NOT respecting our 2 spaces indentation config. Not sure why. Maybe it's due to https://github.com/redhat-developer/vscode-yaml/issues/172.

Currently will remove this extension all together until we find a fix for it.

chiraggshah commented on config/environments/test.rb:
```diff
@@ -10,13 +8,13 @@
-  config.cache_classes = false
-  config.action_view.cache_template_loading = true
+  # Turn false under Spring and add config.action_view.cache_template_loading = true.
+  config.cache_classes = true
```
I agree with you completely. I too have faced issues with spring before. Lets remove it. :+1:

yedhink commented:
@unnitallman @chiraggshah 
**Please don't merge this PR yet. We have to test this out in a neeto repo first.**

I will port neetoEngage to Rails7. That will act as a reference and will help us figure out if any other issues ought to be fixed as part of a neeto repo's upgrade to Rails 7.

We have successfully upgraded Wheel to Rails 7 in this PR. We still use Webpacker underneath the hood. Will take up porting to `esbuild` in a separate issue.

# To be updated in neeto repos
Changes that ought to be ported to all neeto repos:
- Gem version updates - starting from Rails 7 till version upgrades of other gems
- [Uglifier config update](https://github.com/bigbinary/wheel/pull/784/commits/c1869767cada1980ae0a46e125af01a680b123a4) - otherwise Heroku deployment fails
- Removed `spring` gem and its configs all together(`bundle remove spring && git rm -rf bin/spring config/spring.rb`) - https://github.com/bigbinary/wheel/pull/784/commits/9bb3d219ecdb266d04adf63837d98ea68582cc34
-  Added `pry-byebug` gem over separate `pry` and `byebug` gems - https://github.com/bigbinary/wheel/commit/f2c316578552a8079306eec3e61b552798d98df9
- Set config defaults to v7 - it's important to port this - https://github.com/bigbinary/wheel/pull/784/commits/ac428d236effa2e08ce3b53c27a60b03bd3b9fe6
- Ensure exceptions will be raised if i18n translations are missing - https://github.com/bigbinary/wheel/pull/784/commits/3193ed11639198e45dac60001101df9b30edf3c3
- Removed the `listen` gem - Rails 7 takes of this by default - https://github.com/bigbinary/wheel/pull/784/commits/e3eb01c8d5a2534dcd1c10bf5b7f0d5f804dacb1


# Note for all neeto tech leads
Once this PR is merged, all neeto repos will be forced to upgrade to Rails 7. So all of the tech leads can use this PR as reference to understand what all are the changes required. I would only suggest to cherry pick the changes into their corresponding neeto project by going through the file changes in this PR. The reason is that, we have added some of our own configs over the default ones.

yedhink commented:
Facing a blocker issue during upgrade in neeto repos: https://github.com/bigbinary/neeto-area51/issues/9#issue-1142535977

Will have to resolve it before merging this PR.

chiraggshah commented:
# Note for all neeto tech leads (continued...)

Apart from the above changes, each of the repos need to make sure that

Before upgrade:
- Ensure that the test coverage of the application is above 90%. Helps to catch any unwanted bugs introduced during upgrade.
- Fix all the existing deprecation warnings in the logs. This is to ensure that you are up to date and the upgrade to 7.0 would be smoother. Run tests and monitor the test.log file for any deprecation warnings. If the app has sufficient test cases, that should raise all the warnings. But just for being thorough, go through the application features once and monitor the development.log as well. What I generally do is `tail -f ./log/test.log | grep 'DEPRECATION'`. That filters out all other things.

During the upgrade
- Wheel is a barebones app so it would not be utilizing all of rails features. Go through all the changes in https://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#upgrading-from-rails-6-1-to-rails-7-0 and see if your application uses any of those and fix them accordingly. This blog also summarizes the changes nicely https://www.fastruby.io/blog/rails/upgrades/upgrade-rails-6-1-to-7-0.html
- Keep commits incremental. After fixing each type of warning, run the tests and then commit.

After the upgrade
- Try to run the app in production mode on local and fix any leftover errors.

unnitallman commented:
Closing in favor of https://github.com/bigbinary/wheel/pull/874
